File: algo/lc.1.cc
#include <vector>
#include <iostream>
#include <map>

using namespace std;
#include "helper.hpp"

class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        // assume that input is valid
        map<int, int> m;
        map<int, int>::iterator cur;
        for (int i = 0; i < (int)nums.size(); i++) {
            if ((cur = m.find(target-nums[i])) != m.end())
                return vector<int> {i, cur->second};
			m.insert(pair<int,int>(nums[i], i));
        }
        return vector<int>{-1, -1};
    }
};

int
main(void)
{
	auto s = Solution();
	vector<int> v {3, 2, 4};
	cout << s.twoSum(v, 6) << endl;
	return 0;
}

File: algo/lc.1.go
package main

import "fmt"

func twoSum(nums []int, target int) []int {
	m := map[int]int{}
	for k, v := range nums {
		if idx, ok := m[target-v]; !ok {
			m[v] = k
		} else {
			return []int{k, idx}
		}
	}
	return []int{-1, -1}
}

func main() {
	v := []int{3, 2, 4}
	fmt.Println(twoSum(v, 6))
	v = []int{2, 7, 11, 15}
	fmt.Println(twoSum(v, 13))
}
